src/Java.Interop.BootstrapTasks/Java.Interop.BootstrapTasks/JdkInfo.cs

```diff
@@ -14,74 +17,265 @@ public class JdkInfo : Task
+				javacPath   = Path.Combine (java_home, "bin", "javac");
+				if (OnMacOS) {
+					jdkJvmPath  = Directory.EnumerateFiles (java_home, "libjli.dylib", SearchOption.AllDirectories)
+						.FirstOrDefault ();
+				} else {
+					jdkJvmPath  = Path.Combine (java_home, "jre", "bin", "server", "libjvm.so");
+				}
+			}
 
-			var jvmPath = Path.Combine (javaSdkPath, "jre", "bin", "server", "jvm.dll");
-			if (!File.Exists (jvmPath)) {
-				Log.LogError ($"JdkJvmPath not found at {jvmPath}");
+			if (!File.Exists (jarPath)) {
+				Log.LogError ("Could not determine location `jar`.");
+				return false;
+			}
+			if (!File.Exists (javacPath)) {
+				Log.LogError ("Could not determine location `javac`.");
+				return false;
+			}
+			if (string.IsNullOrEmpty (jdkJvmPath)) {
+				Log.LogError ("Could not determine location of JDK embedding library, e.g. `JVM.DLL`.");
 				return false;
 			}
 
-			var javaIncludePath = Path.Combine (javaSdkPath, "include");
-			var includes = new List<string> { javaIncludePath };
-			includes.AddRange (Directory.GetDirectories (javaIncludePath)); //Include dirs such as "win32"
-
-			var includeXmlTags = new StringBuilder ();
-			foreach (var include in includes) {
-				includeXmlTags.AppendLine ($"<JdkIncludePath Include=\"{include}\" />");
-			}
-
-			Directory.CreateDirectory (Path.GetDirectoryName (Output.ItemSpec));
-			File.WriteAllText (Output.ItemSpec, $@"<Project xmlns=""http://schemas.microsoft.com/developer/msbuild/2003"">
-  <Choose>
-    <When Condition="" '$(JdkJvmPath)' == '' "">
-      <PropertyGroup>
-        <JdkJvmPath>{jvmPath}</JdkJvmPath>
-      </PropertyGroup>
-      <ItemGroup>
-        {includeXmlTags}
-      </ItemGroup>
-    </When>
-  </Choose>
-  <PropertyGroup>
-    <JavaCPath Condition="" '$(JavaCPath)' == '' "">{Path.Combine (javaSdkPath, "bin", "javac.exe")}</JavaCPath>
-    <JarPath Condition="" '$(JarPath)' == '' "">{Path.Combine (javaSdkPath, "bin", "jar.exe")}</JarPath>
-  </PropertyGroup>
-</Project>");
-
-			JavaSdkDirectory = javaSdkPath;
-			Log.LogMessage (MessageImportance.Low, $"  [Output] {nameof (JavaSdkDirectory)}: {JavaSdkDirectory}");
+			var javaIncludePath = Path.Combine (java_home, "include");
+			includes.Add (javaIncludePath);
+			includes.AddRange (Directory.GetDirectories (javaIncludePath));
+
+			JavaHomePath  = java_home;
+
+			Directory.CreateDirectory (Path.GetDirectoryName (PropertyFile.ItemSpec));
+			Directory.CreateDirectory (Path.GetDirectoryName (MakeFragmentFile.ItemSpec));
+
+			WritePropertyFile (jarPath, javacPath, jdkJvmPath, includes);
+			WriteMakeFragmentFile (jarPath, javacPath, jdkJvmPath, includes);
+
+			Log.LogMessage (MessageImportance.Low, $"  [Output] {nameof (JavaHomePath)}: {JavaHomePath}");
 
 			return !Log.HasLoggedErrors;
 		}
 
-		string GetJavaSdkPath ()
+		Version GetMaxJdkVersion ()
+		{
+			if (string.IsNullOrEmpty (MaximumJdkVersion))
+				return null;
+			if (!MaximumJdkVersion.Contains (".")) {
```

Comment:
`MaximumJdkVersion.IndexOf (".", StringComparison.Ordinal)` is going to be faster

Reply:
`Contains` is clearer, certainly, but I think we should strike balance between clarity and performance - too many tiny sacrifices of the latter add up. Granted, this particular code will run during `make preapre` but I think it's a good idea to "promote" code that preforms...